test/box-luatest/gh_9643_priv_origins_test.lua

```diff
@@ -0,0 +1,182 @@
+        t.assert_equals(err.message, exp_err)
+        t.assert_equals(_priv:get({uid, 'space', sid}), exp)
+
+        exp[5] = bit.bor(box.priv.D, box.priv.R)
+        exp[6] = {origins = {
+            [''] = bit.bor(box.priv.D, box.priv.R),
+            one = box.priv.D,
+        }}
+        revoke('myuser', 'write', 'space', 'myspace', {_origin = 'one'})
+        t.assert_equals(_priv:get({uid, 'space', sid}), exp)
+
+        exp[5] = box.priv.D
+        exp[6] = {origins = {
+            [''] = box.priv.D,
+            one = box.priv.D,
+        }}
+        revoke('myuser', 'read', 'space', 'myspace')
+        t.assert_equals(_priv:get({uid, 'space', sid}), exp)
+
+        exp[5] = box.priv.D
+        exp[6] = nil
+        revoke('myuser', 'drop', 'space', 'myspace', {_origin = 'one'})
+        t.assert_equals(_priv:get({uid, 'space', sid}), exp)
+
+        revoke('myuser', 'drop', 'space', 'myspace')
+        t.assert_equals(_priv:get({uid, 'space', sid}), nil)
+    end)
+end
+
+g.test_origins_in_roles = function(cg)
+    cg.server:exec(function()
+        box.schema.role.create('myrole')
+        box.schema.role.create('somerole')
+        local rid = box.space._user.index.name:get('myrole').id
+        local tid = box.space._user.index.name:get('somerole').id
+        local grant = box.schema.role.grant
+        local revoke = box.schema.role.revoke
+        local _priv = box.space._priv
+
+        local exp = {0, rid, 'role', tid, 0}
+        t.assert_equals(_priv:get({rid, 'role', tid}), nil)
+
+        -- Check grant with origins.
+        exp[5] = box.priv.X
+        exp[6] = nil
+        grant('myrole', 'somerole', nil, nil)
+        t.assert_equals(_priv:get({rid, 'role', tid}), exp)
+
+        exp[6] = {origins = {
+            [''] = box.priv.X,
+            two = box.priv.X,
+        }}
+        grant('myrole', 'somerole', nil, nil, {_origin = 'two'})
+        t.assert_equals(_priv:get({rid, 'role', tid}), exp)
+
+        local ok, err = pcall(grant, 'myrole', 'somerole')
+        local exp_err = "User 'myrole' already has role 'somerole'"
+        t.assert(not ok)
+        t.assert_equals(err.message, exp_err)
+        t.assert_equals(_priv:get({rid, 'role', tid}), exp)
+
+        ok, err = pcall(grant, 'myrole', 'somerole', nil, nil,
+                        {_origin = 'two'})
+        t.assert(not ok)
+        t.assert_equals(err.message, exp_err)
+        t.assert_equals(_priv:get({rid, 'role', tid}), exp)
+
+        -- Check revoke with origins.
+        exp[6] = {origins = {
+            two = box.priv.X,
+        }}
+        revoke('myrole', 'somerole')
+        t.assert_equals(_priv:get({rid, 'role', tid}), exp)
+
+        ok, err = pcall(revoke, 'myrole', 'somerole')
+        exp_err = "Role 'myrole' does not have role 'somerole' provided by " ..
+                  "default origin"
+        t.assert(not ok)
+        t.assert_equals(err.message, exp_err)
+        t.assert_equals(_priv:get({rid, 'role', tid}), exp)
```

Comment:
Please also test that `if_exists` and `if_not_exists` options work with origins.

Reply:
Thanks, added a few tests.